Preview future site state

PatelUtkarsh opened the pull request:
Fixes #21 

**Todos**
- [x] Add feature to preview site state
- [x] Add unit-test 
- [x] Add WP date-picker 
- [x] Update unit-test for date picker
- [x] Update css for date picker
- [x] Improve code coverage

Dependent on PR #93

PatelUtkarsh committed:
Add dashboard widget for preview future scheduled snapshots #21

PatelUtkarsh committed:
Add unit-test for preview future snapshot

PatelUtkarsh committed:
Add default style WP date picker for dashboard widget

PatelUtkarsh committed:
Add error code for preview scheduled snapshots

PatelUtkarsh committed:
Add dashboard css for preview future site state widget

PatelUtkarsh committed:
Handle case on where there is only one future snapshot

PatelUtkarsh committed:
Update dashboard widget unit-tests

PatelUtkarsh committed:
Remove datepicker todo

PatelUtkarsh committed:
Update unit-test

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

miina commented:
@PatelUtkarsh Bringing the conversation over to here:

> M: When checking the future preview with 1 Snapshot (without conflicts), I could still edit it in Customizer and change a value with clicking "Schedule", is this expected behavior? Maybe it is :)
> 
> U: I haven't got approval on that one yet, So if there is only one snapshot we are directly previewing it right now in (#95), Maybe we should fork/duplicate it and preview it. I thought because it is just previewing we can skip extra steps of duplication.
> 
> M: When checking the future preview with multiple Snapshots scheduled and when I click Customize and make some edits, it's going to create a new merged Snapshot, is this expected? However, this new Snapshot will not be Scheduled unless scheduled specifically, so it won't actually make any changes. Maybe this can be confusing? Thoughts?
> 
> U: Yes, so when you preview from dashboard it creates new merge snapshot with auto-draft status(merging all future snapshots comparing to date). If you start editing it, it will put that in draft status and will allow you to edit.

Okay, thinking that it would be great if previewing a future site would have consistency. So that if I make edits in Customizer while previewing a future site, always X happens. Otherwise it's confusing in terms of UX that in case in code only a single Snapshot is found, then X happens but if more Snapshots are found, Y happens.
Also, it would be good to know somehow that I'm not previewing a Snapshot I've created but the site preview. For example when live previewing a theme, the Customizer has this button instead of the usual `Save & Publish`:
![live-preview](https://cloud.githubusercontent.com/assets/3294597/19725960/f24ee95c-9b5e-11e6-929a-87ebfb91b283.png)
Maybe when making edits while previewing, it'd also be useful to use a similar logic and change the button text? Thoughts?

PatelUtkarsh commented:
So in a case of a single snapshot we should create a duplicate snapshot to remove that inconsistency (depends on #90), I will add to-do.

About UX I am not sure, @westonruter thoughts?

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

westonruter commented:
Should previewing future state where multiple changesets/snapshots are combined have a readonly mode? Should previewing this open on the frontend with the snapshot UUID supplied rather than opening the customizer with the UUID supplied? I should think that the virtual future-merged changesets would be represented in a new `auto-draft` snapshot/changeset that would be readonly and wouldn't be used for making edits. If you did want to explicitly edit the merged state, then you could click multiple future snapshots to then merge together into a new snapshot to make that explicit. Some ideas.

PatelUtkarsh committed:
Merge branch 'develop' into feature/preview-future-site-state

PatelUtkarsh committed:
Remove inconsistency on preview feature state,
Remove inspect snapshot and update customize url on preview of future site state from admin bar

PatelUtkarsh committed:
Add test-case when there is only 1 snapshot to preview in future state

PatelUtkarsh commented:
Yes, read-only makes sense.
In the case of the read-only preview, I've removed wp_admin_bar link for `inspect snapshot`(post-edit) link and `customize` (don't add UUID in this case).  I've done that in 1e043e9.

PatelUtkarsh committed:
Remove customize link from admin-bar when previewing future site

PatelUtkarsh committed:
Add post meta to handle if post is to preview future site state

PatelUtkarsh committed:
Merge 'feature/resolve-conflict' into 'feature/preview-future-site-state'

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

 Conflicts:
	php/class-customize-snapshot-manager.php
	tests/php/test-class-customize-snapshot-manager.php

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Fix phpcs

valendesigns commented:
@PatelUtkarsh Please fix merge conflicts so I can review. Thanks!

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh committed:
Merge branch 'feature/resolve-conflict' into feature/preview-future-site-state

PatelUtkarsh commented:
Todo:
- [x] Remove publish changeset button from admin bar when previewing future state.
- [x] Test

PatelUtkarsh committed:
Fix phpcs

PatelUtkarsh committed:
Fix admin bar when previewing future state of site

PatelUtkarsh committed:
Remove unnecessary diff

PatelUtkarsh committed:
Fix error, Merge by post_date for schedule status

PatelUtkarsh commented on php/class-post-type.php:
```diff
@@ -928,15 +963,24 @@ public function handle_snapshot_merge( $redirect_to, $do_action, $post_ids ) {
+				$compare_key_b = 'post_date';
+			}
+			$compare_a = $a->$compare_key_a;
+			$compare_b = $b->$compare_key_b;
```
@westonruter When you review this PR, I've made this change on snapshot merge function. 
This will give priority to scheduled post on merge. This was needed for preview future site state but this will also now affect changeset merge option.

Let know your thoughts, earlier we were merging based on `post_modified` without checking status.
Now in case of scheduled post it uses `post_date` otherwise it will use `post_modified`.

westonruter replied:
This makes sense to me 👍

PatelUtkarsh committed:
Fix unit-test

westonruter commented on php/class-customize-snapshot-manager.php:
```diff
@@ -386,11 +394,26 @@ static public function is_valid_uuid( $uuid ) {
+		// If previewing future state of changeset only add exit changeset preview link in admin bar.
+		$wp_customize = $this->get_customize_manager();
+		$is_future_state_preview = (
+				is_customize_preview()
```
Looks like too much indentation here.

westonruter commented on php/class-customize-snapshot-manager.php:
```diff
@@ -576,9 +599,14 @@ public function add_snapshot_exit_link( $wp_admin_bar ) {
+		$wp_customize = $this->get_customize_manager();
+		$title = __( 'Exit Changeset Preview', 'customize-snapshots' );
+		if ( $this->is_previewing_future_state_changeset( $wp_customize->changeset_post_id() ) ) {
+			$title = __( 'Exit Preview Of Future Site State', 'customize-snapshots' );
```
“Of” should be just “of” according to English style.

westonruter commented on php/class-dashboard-widget.php:
```diff
@@ -0,0 +1,200 @@
+		$date_time = current_time( 'mysql' );
+		$date_time = new \DateTime( $date_time );
+		if ( isset( $_POST['year'], $_POST['month'], $_POST['day'], $_POST['hour'], $_POST['minute'], $_REQUEST['_wpnonce'] ) && wp_verify_nonce( $_REQUEST['_wpnonce'], 'customize_site_state_future_snapshot_preview' ) ) {
+			$date_time->setTimestamp( strtotime( "{$_POST['year']}-{$_POST['month']}-{$_POST['day']} {$_POST['hour']}:{$_POST['minute']}" ) );
```
These should probably be validated as only containing integers.

westonruter commented on php/class-dashboard-widget.php:
```diff
@@ -0,0 +1,200 @@
+						<?php $month = Customize_Snapshot_Manager_Compat::get_month_choices(); ?>
+						<select id="snapshot-date-month" class="date-input month" data-date-input="month" name="month">
+							<?php
+							foreach ( $month['month_choices'] as $month_choice ) {
```
Since we're embedding PHP in HTML, let's use the alternative control structure: 

```php
<?php foreach ( $month['month_choices'] as $month_choice ) : ?>
```

And then do:

```php
<option value="<?php echo esc_attr( ... ) ?>...
```

Instead of `echo`'ing a string.

westonruter commented on php/class-dashboard-widget.php:
```diff
@@ -0,0 +1,200 @@
+			echo '<p class="error-message">' . esc_html( $this->error[ $this->error_code ] ) . '</p>';
+		} ?>
+		<form method="post">
+				<div class="preview-future-state date-inputs clear">
```
Looks like this is indented one too many tabs.

westonruter commented on php/class-dashboard-widget.php:
```diff
@@ -0,0 +1,200 @@
+			return;
+		}
+		$date = new \DateTime();
+		$date->setTimestamp( strtotime( "{$_POST['year']}-{$_POST['month']}-{$_POST['day']} {$_POST['hour']}:{$_POST['minute']}" ) );
```
Probably should validate all these vars are digits.

westonruter commented on php/class-post-type.php:
```diff
@@ -395,6 +396,34 @@ public function render_data_metabox( $post ) {
+						echo get_the_title( $uuid_post ) . ' - ' . esc_html( $uuid_post->post_name );
+					}
+				}
+				echo ( 'future' === $uuid_post->post_status ) ? ' (Scheduled)' : '';
```
Lacking translation for `(Scheduled)`.

PatelUtkarsh replied:
Part of #93 will fix in it.

westonruter commented on php/class-post-type.php:
```diff
@@ -970,22 +1014,56 @@ public function merge_snapshots( $post_ids ) {
+					$uuid[] = $post_data['uuid'];
+				}
+			}
+			$merged_snapshot_data[ $key ]['merged_uuid'] = array_values( array_unique( $uuid ) );
```
Instead of `merged_uuid` what about `merged_changeset_uuids` or `merged_changesets`?

PatelUtkarsh replied:
Will use `merged_changeset_uuids` part of #93 I'll fix in it.

westonruter commented on php/class-post-type.php:
```diff
@@ -970,22 +1014,56 @@ public function merge_snapshots( $post_ids ) {
+				}
+			}
+			$values = wp_list_pluck( $original_values, 'value' );
+			$values = array_unique( $values, SORT_REGULAR );
```
Nice, I didn't know about the second arg for `array_unique()`.

westonruter commented on php/class-post-type.php:
```diff
@@ -1331,4 +1409,134 @@ public function handle_snapshot_fork() {
+	 */
+	public function resolve_conflict_markup( $setting_id, $value, $snapshot_content, $post ) {
+		if ( isset( $value['merge_conflict'] ) ) {
+			$setting_id_key = str_replace( ']', '\\]', str_replace( '[', '\\[', $setting_id ) );
```
Why this? Why adding escapes?

PatelUtkarsh replied:
@westonruter 
`#TB_inline?width=600&height=550&inlineId=snapshot-post\[post\]\[1\]`
Thickbox uses jQuery, `inlineId` is selector and for `[` to work in jQuery it needs this escaping https://stackoverflow.com/a/2365005

PatelUtkarsh replied:
https://api.jquery.com/category/selectors/
>To use any of the meta-characters ( such as  !"#$%&'()*+,./:;<=>?@[\]^`{|}~ ) as a literal part of a name, it must be escaped with with two backslashes: \\.

westonruter commented on php/class-post-type.php:
```diff
@@ -1331,4 +1409,134 @@ public function handle_snapshot_fork() {
+	 * Change conflicted value in post_content, if they were changed in the meta box.
+	 *
+	 * @param String $content Post content to filter.
+	 * @return String $content Post content, possibly filtered.
```
Add `@global \WP_Post $post`

PatelUtkarsh replied:
Will add in #93.

westonruter commented on php/class-post-type.php:
```diff
@@ -1331,4 +1409,134 @@ public function handle_snapshot_fork() {
+				}
+			}
+		}
+		$content = Customize_Snapshot_Manager::encode_json( $snapshot_data );
```
Shouldn't this re-`wp_slash()`?

PatelUtkarsh replied:
Yes.

westonruter commented on php/class-post-type.php:
```diff
@@ -1331,4 +1409,134 @@ public function handle_snapshot_fork() {
+	 * @param array $post_content snapshot data.
+	 * @return array merged snapshot.
+	 */
+	public function get_snapshot_merged_uuid( $post_content ) {
```
Maybe this should be `get_merged_changesets()`. Let's also make it `private` as I don't see a need for it to be exposed.

PatelUtkarsh replied:
Yep, will update in #93.

PatelUtkarsh replied:
```php
	/**
	 * Get snapshot merged uuid from post_content
	 *
	 * @internal
	 *
	 * @param array $post_content snapshot data.
	 * @return array merged snapshot.
	 */
	public function _get_merged_changesets( $post_content ) {
```
@westonruter How about this? 
If we make this private we need to mock that method in unit-test which is not best practice. We can cover this method under render function's unit-test.

PatelUtkarsh committed:
Code review changes

westonruter committed:
Let dashboard widget's contents vary according to number of scheduled changests.

When no changesets are scheduled, show message explaining the metabox but then not showing the form. When there is only one changeset scheduled, show permalink for just that one changes.

Use an error notice when an error happens after submitting the form. Use autofocus to scroll form back into view after reloading page.

westonruter committed:
Merge branch 'develop' of https://github.com/xwp/wp-customize-snapshots into feature/preview-future-site-state
